unified-runtime/source/adapters/opencl/context.cpp

```diff
@@ -111,51 +122,47 @@ urContextRelease(ur_context_handle_t hContext) {
   // should drastically reduce the chances of the pathological case described
   // in the comments in common.hpp.
   static std::mutex contextReleaseMutex;
-  auto clContext = cl_adapter::cast<cl_context>(hContext);
-
-  {
-    std::lock_guard<std::mutex> lock(contextReleaseMutex);
-    size_t refCount = 0;
-    CL_RETURN_ON_FAILURE(clGetContextInfo(clContext, CL_CONTEXT_REFERENCE_COUNT,
-                                          sizeof(size_t), &refCount, nullptr));
-
-    // ExtFuncPtrCache is destroyed in an atexit() callback, so it doesn't
-    // necessarily outlive the adapter (or all the contexts).
-    if (refCount == 1 && cl_ext::ExtFuncPtrCache) {
-      cl_ext::ExtFuncPtrCache->clearCache(clContext);
-    }
+  auto clContext = hContext->CLContext;
+
+  std::lock_guard<std::mutex> lock(contextReleaseMutex);
+  size_t refCount = hContext->getReferenceCount();
+  // ExtFuncPtrCache is destroyed in an atexit() callback, so it doesn't
+  // necessarily outlive the adapter (or all the contexts).
+  if (refCount == 1 && cl_ext::ExtFuncPtrCache) {
+    cl_ext::ExtFuncPtrCache->clearCache(clContext);
   }
```

Comment:
could do fewer atomic accesses here
```suggestion
  if (hContext->decrementReferenceCount() == 0 ) {
    // ExtFuncPtrCache is destroyed in an atexit() callback, so it doesn't
    // necessarily outlive the adapter (or all the contexts).
    if(cl_ext::ExtFuncPtrCache) {
      cl_ext::ExtFuncPtrCache->clearCache(clContext);
    }
    delete hContext;
  }
```